text/0048-schema-annotations.md

```diff
@@ -0,0 +1,140 @@
+# Annotations for Schemas 
+
+## Related issues and PRs
+
+- Reference Issues: 
+- Implementation PR(s): 
+
+## Timeline
+
+- Started: 2024-02-05
+
+## Summary
+
+Like Cedar policies, users may want to associate arbitrary, machine readable metadata with Schema objects.
+We solved this problem in Cedar Policies by allowing for *annotations*: arbitrary key/value pairs that are attachable to policies.
+This could be extended to Cedar Schemas, allowing users to attach attributes an entity type, common type, and action definitions.
+
+
+## Basic example
+
+1. Doc comments
+```
+@doc("This entity defines our central user type")
+entity User { 
+    owner : User,
+    team : String
+}
+```
+
+```
+@doc("Stop users from accessing a high security documents unless they are in the same building")
+forbid(principal, action, resource) when {
+    resource.security_level == "HIGH"
+unless {
+    resource.location == principal.location
+};
+```
+
+## Motivation
+
+Users should be allowed to associate machine readable metadata with objects in a Schema.
+While we could create special syntax for associating particular kinds of metadata, we cannot
+predict all of the metadata uses that users will have. 
+Thus providing a flexible system that users can adapt to their needs is preferrable.
+This proposal re-uses the same syntax from Cedar Policies, creating a unified syntax.
+
+
+## Detailed design
+
+### Semantics
+Attributes have **no** impact on validation decisions. 
+Attributes are arbitrary key/value pairs where:
+* 'key' is a valid Cedar identifier
+* 'value' is a Cedar string
+
+
+The Cedar spec takes no opinion or stance on the interpretation of annotations.
+The interpretation is entirely up to users of Cedar.
+
+### Human Readable Syntax
+Attributes in Cedar Schemas will mirror the syntax used for attributes in a policy: informally that's `@<key>("value")`.
+Formally the following rule is added to the Cedar grammar: 
+```
+Annotation := '@' IDENT '(' STR ')'
+Annotations := Annotation {Annotations}
+```
+With an arbitrary number of them being able to prepend to a top level declaration in a schema.
+
+Thus the full schema syntax becomes:
+```
+Schema      := {Namespace}
+Namespace   := ('namespace' Path '{' {Decl} '}') | {Decl}
+Decl        := Entity | Action | TypeDecl
+Entity      := Annotations 'entity' Idents ['in' EntOrTyps] [['='] RecType] ';'
+Action      := Annotations 'action' Names ['in' (Name | '[' [Names] ']')] [AppliesTo] [ActAttrs]';'
```

Comment:
Annotations on actions could be a bit confusing when annotations aren't reflected in any way at run time. It kind of feels like a second set of pseudo-attributes.

Reply:
I agree that I think annotations on entity types and common types are more clear than annotations on actions.

Reply:
For the doc-comments use-case, you want annotations pretty much everywhere, e.g., to describe what actions do, right?

Reply:
That is true.  As noted in a different thread, I think we want to allow annotations (particularly doc comments) on individual attributes of entity/common types as well.

Reply:
We probably do want annotations basically everywhere. It's _possibly_ confusing on actions, but leaving it out is guaranteed to annoy users.